nifi-nar-bundles/nifi-standard-services/nifi-record-serialization-services-bundle/nifi-record-serialization-services/src/main/java/org/apache/nifi/cef/CEFRecordReader.java

```diff
@@ -0,0 +1,216 @@
+        final boolean acceptEmptyExtensions
+    ) {
+        this.reader = new BufferedReader(new InputStreamReader(inputStream));
+        this.schema = recordSchema;
+        this.parser = parser;
+        this.logger = logger;
+        this.locale = locale;
+        this.rawMessageField = rawMessageField;
+        this.invalidField = invalidField;
+        this.includeCustomExtensions = includeCustomExtensions;
+        this.acceptEmptyExtensions = acceptEmptyExtensions;
+    }
+
+    @Override
+    public Record nextRecord(final boolean coerceTypes, final boolean dropUnknownFields) throws IOException, MalformedRecordException {
+        final String line = nextLine();
+
+        if (line == null) {
+            return null;
+        }
+
+        final CommonEvent event = parser.parse(line, true, acceptEmptyExtensions, locale);
+
+        if (event == null) {
+            logger.debug("Event parsing resulted no event");
+
+            if (invalidField != null && !invalidField.isEmpty()) {
+                return new MapRecord(schema, Collections.singletonMap(invalidField, line));
+            } else {
+                throw new MalformedRecordException("The following line could not be parsed by the CEF parser: " + line);
+            }
+        }
+
+        final Map<String, Object> values = new HashMap<>();
+
+        try {
+            event.getHeader().entrySet().forEach(field -> values.put(field.getKey(), convertValue(field.getKey(), field.getValue(), coerceTypes)));
+            event.getExtension(true, includeCustomExtensions).entrySet().forEach(field -> values.put(field.getKey(), convertValue(field.getKey() ,field.getValue(), coerceTypes)));
+
+            for (final String fieldName : schema.getFieldNames()) {
+                if (!values.containsKey(fieldName)) {
+                    values.put(fieldName, null);
+                }
+            }
+
+        } catch (final CEFHandlingException e) {
+            throw new MalformedRecordException("Error during extracting information from CEF event", e);
+        }
+
+        if (rawMessageField != null) {
+            values.put(rawMessageField, line);
+        }
+
+        return new MapRecord(schema, values, true, dropUnknownFields);
+    }
+
+    private String nextLine() throws IOException {
+        String line;
+
+        while((line = reader.readLine()) != null) {
+            if (!line.isEmpty()) {
+                break;
+            }
+        }
+
+        return line;
+    }
+
+    private Object convertValue(final String fieldName, final Object fieldValue, final boolean coerceType) {
+        final DataType dataType = schema.getDataType(fieldName).orElse(RecordFieldType.STRING.getDataType());
+
+        return coerceType
+            ? convert(fieldValue, dataType, fieldName)
+            : convertSimpleIfPossible(fieldValue, dataType, fieldName);
+    }
+
+    private final Object convert(final Object value, final DataType dataType, final String fieldName) {
```

Comment:
The `final` keyword is unnecessary on private methods.
```suggestion
    private Object convert(final Object value, final DataType dataType, final String fieldName) {
```